src/miniscript/decode.rs

```diff
@@ -66,28 +72,70 @@ macro_rules! match_token {
     };
 }
 
-fn rewrap1<Pk, Pkh, F>(term: &mut Vec<AstElem<Pk, Pkh>>, wrap: F)
-where
-    F: FnOnce(Box<AstElem<Pk, Pkh>>) -> AstElem<Pk, Pkh>
-{
-    let top = term.pop().unwrap();
-    term.push(wrap(Box::new(top)));
+trait Decodable <Pk, Pkh>{
+    fn push_ast0(&mut self, ms : AstElem<Pk, Pkh>) -> Result<(), Error>
```

Comment:
I think it'd be a bit cleaner to write a new `TerminalStack<Pk,Pkh>(Vec<Miniscript<Pk,Pkh>>)` type and implement these methods on them, rather than using a trait.

I also think they should be called `reduce0`, `reduce1`, `reduce2` rather than `push_` because they do more than just push.